File: trees & graphs/C++/binary_search_tree.cc
/**
 * @file binary_search_tree.cc
 * @author Hongyu Ding 
 * @brief This is an example of Binary Search Tree
 * @version 0.1
 * @date 2020-11-27
 * 
 * @copyright Copyright (c) 2020
 * 
 */

#include <iostream>
#include <memory>
#include <vector>

struct Node {
    int key;
    std::shared_ptr<Node> left;
    std::shared_ptr<Node> right;

    Node(int key) : key{key} {}
};

class BST {
    std::shared_ptr<Node> root;

    void print(std::ostream &out, std::shared_ptr<Node> node, int depth) {
        if (!node) return;
        for (int i = 0; i < depth; i ++) {
            out << "    ";
        }
        out << node->key << std::endl;
        print(out, node->left, depth + 1);
        print(out, node->right, depth + 1);
    }

    void inorder(std::ostream &out, std::shared_ptr<Node> node) {
        if (!node) return;
        inorder(out, node->left);
        out << node->key << " ";
        inorder(out, node->right);
    }

    std::shared_ptr<Node> search(std::shared_ptr<Node> node, int key) {
        if (!node) return nullptr;
        if (node->key == key) return node;
        if (node->key < key) {
            return search(node->right, key);
        } else {
            return search(node->left, key);
        }
        return node;
    }

    std::shared_ptr<Node> insert(std::shared_ptr<Node> node, int key) {
        if (!node) return std::make_shared<Node>(key);
        if (node->key == key) return node;
        if (node->key < key) {
            node->right = insert(node->right, key);
        } else {
            node->left = insert(node->left, key);
        }
        return node;
    }

    std::shared_ptr<Node> getMin(std::shared_ptr<Node> node) {
        std::shared_ptr<Node> min = node;
        while (min && min->left) {
            min = min->left;
        }
        return min;
    }

    std::shared_ptr<Node> remove(std::shared_ptr<Node> node, int key) {
        if (!node) return nullptr;
        if (node->key == key) {
            if (!node->left) {
                std::shared_ptr<Node> temp = node->right;
                node.reset();
                return temp;
            } else if (!node->right) {
                std::shared_ptr<Node> temp = node->left;
                node.reset();
                return temp;
            }
            std::shared_ptr<Node> temp = getMin(node->right);
            node->key = temp->key;
            node->right = remove(node->right, temp->key);
        } else if (node->key < key) {
            node->right = remove(node->right, key);
        } else if (node->key > key) {
            node->left = remove(node->left, key);
        }
        return node;
    }

public:
    BST() : root{nullptr} {}
    BST(std::vector<int> &ary) {
        for (auto val : ary) {
            insert(val);
        }
    }

    void print(std::ostream &out) {
        print(out, root, 0);
    }

    void inorder(std::ostream &out) {
        inorder(out, root);
    }

    std::shared_ptr<Node> search(int key) {
        return search(root, key);
    }

    void insert(int key) {
        if (root) {
            insert(root, key);
        } else {
            root = insert(root, key);
        }
    }

    void remove(int key) {
        remove(root, key);
    }
};

int main() {
    BST tree = BST();
    char command;
    while (std::cin >> command) {
        switch (command) {
            case 'a': {
                int val;
                std::cin >> val;
                tree.insert(val);
                break;
            }
            case 'd': {
                int val;
                std::cin >> val;
                tree.remove(val);
                break;
            }
            case 'p': {
                tree.print(std::cout);
                break;
            }
            case 's': {
                int val;
                std::cin >> val;
                std::cout << tree.search(val)->key << std::endl;
                break;
            }
            case 'i': {
                tree.inorder(std::cout);
                std::cout << std::endl;
            }
        }
    }
    return 0;
}

File: trees & graphs/Java/BinarySearchTree.java
/**
 * @File BinarySearchTree.java
 * @Author Weilan Tao
 * @Date 2020-12-06
 * @Description This is the implementation for Binary Search Tree, Java version.
 * @Since version-1.2
 * @Copyright Copyright (c) 2020
 */


public class BinarySearchTree {
    private Node root;

    class Node{
        int value;
        Node left;
        Node right;

        Node (Integer value){
            this.value=value;
            right=null;
            left=null;
        }
    }
    public  BinarySearchTree(){

    }

    public void insert(int val ){
        root =insertRec(root, val);
    }

    private Node insertRec(Node node, int val){
        if(node==null){
            return new Node(val);
        }

        if(node.value==val){
            return node;
        }else if(node.value<val){
            node.right=insertRec(node.right,val);
        }else{
            node.left=insertRec(node.left,val);
        }

        return node;
    }



    public void printTree(){
        if(root!=null){
            printTreeREC(root.right,"    ",true);
            System.out.println("+---"+root.value);
            printTreeREC(root.left,"    ",false);
        }

    }

    private void printTreeREC(Node node, String pfx,  boolean isRight){
        if(node==null){
            return;
        }

        printTreeREC(node.right,pfx+(isRight ? "    " : "|   "),true);
        System.out.println(pfx+"+---"+node.value);
        printTreeREC(node.left,pfx+(isRight ? "|   " : "    "),false);
    }

    public void inOrder(){
        inOrderRec(root);
        System.out.println(" ");
    }
    private void inOrderRec(Node node){
        if (node!=null){
            inOrderRec(node.left);
            System.out.print(node.value+" ");
            inOrderRec(node.right);
        }
    }

    public boolean contains(int val){
        return  searchRec(root, val)!=null;
    }

    private Node searchRec(Node node, int val){
        if(node==null){
            return null;
        }else{
            if(node.value==val){
                return node;
            }
            else if(node.value>val){
                return searchRec(node.left,val);
            }else if(node.value<val){
                return searchRec(node.right,val);
            }
            return node;
        }
    }

    public Node getMin(Node node) {
        if (node == null) {
            return node;
        } else {
            while (node.left != null) {
                node = node.left;
            }
            return node;
        }
    }

    public Node remove(Node node, int value){
        if(node==null){
        }else{
            if (node.value==value){
                //have two child nodes
                if(node.left!=null && node.right!=null){
                    Node temp=getMin(node.right);
                    node.value=temp.value;
                    node.right=remove(node.right,temp.value);
                    temp=null;
                }
                else if(node.left!=null){
                    return node.left;
                }else if(node.right!=null){

                    return node.right;
                }
                else {
                    node=null;
                    return node;
                }

            }else if(node.value>value){
                node.left=remove(node.left, value);
            }else if (node.value<value){
                node.right=remove(node.right, value);
            }
        }
        return node;
    }

    public void remove(int value ){
        if(contains(value)) remove(root, value);
    }
}

class TestBST{
    public static void main(String [] args){
        BinarySearchTree bst = new BinarySearchTree();
        bst.insert(50);
        bst.insert(30);
        bst.insert(80);
        bst.insert(20);
        bst.insert(35);
        bst.insert(34);
        bst.insert(32);
        bst.insert(40);
        bst.insert(70);
        bst.insert(75);
        bst.insert(100);

        bst.printTree();
        bst.inOrder();
        System.out.println(bst.contains(4));

        bst.remove(70);
        bst.remove(4);

        bst.printTree();

        bst.remove(30);
        bst.printTree();


    }
}
